lib/smooth/application/display/DisplaySpi.cpp

```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
-    // for the same rotation.  For example M5Stack sets madctl to 0x68 for
-    // portrait while the adafruit 2.88 dipslay sets madctl to 0x48.  To make
-    // driver more versatile I have used madctl_value instead of the typical
-    // 0-3 screen_rotation_value that is used in switch statement.
+    // Most display ICs use MADCTL to set screen orientation. Typically there
+    // are four different screen orientations - Portrait, Landsacpe,
+    // Portrait-Flipped and Landscape-Flipped.  The problem is there is not
+    // a universal standard for setting the MADCTL value for a particular
+    // screen oreintation.  For example M5Stack uses a MADCT value of 0x68
+    // for portrait while the Adafruit 2.88 dipslay uses a MADCTL value of 0x48.
+    // To make driver more versatile the function set_madctl is used to allow user
+    // to set the screen orientation instead of the typical 0-3
+    // screen_rotation_constant that is used in switch statement.
     // If you are using LittlevGL then you need to set the appropriate screen
     // width and screen height for the rotation you have chosen in the lvconf.h file.
-    bool ILI9341::set_rotation(uint8_t madctl_value)
+    bool DisplaySpi::set_madctl(uint8_t value, uint8_t madctl_reg)
     {
-        bool res = lcd_wr_cmd(0x36);
-        res &= lcd_wr_data(&madctl_value, 1);
+        bool res = send_cmd(madctl_reg);
+        res &= send_data(&value, 1);
 
         return res;
     }
 
-    // Reset the display
-    void ILI9341::reset_display()
+    // Hardware Reset
+    void DisplaySpi::hw_reset(bool active_low, int active_time, int delay_time)
     {
-        reset_pin.set(PIN_LOW);
-        std::this_thread::sleep_for(std::chrono::milliseconds(100));
-        reset_pin.set(PIN_HIGH);
-        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        if (active_low)
+        {
+            reset_pin.set(PIN_LOW);     // force the display chip to reset
+            std::this_thread::sleep_for(std::chrono::milliseconds(active_time));
+            reset_pin.set(PIN_HIGH);
+            std::this_thread::sleep_for(std::chrono::milliseconds(delay_time));
+        }
+        else
+        {
+            reset_pin.set(PIN_HIGH);    // force the display chip to reset
+            std::this_thread::sleep_for(std::chrono::milliseconds(active_time));
+            reset_pin.set(PIN_LOW);
+            std::this_thread::sleep_for(std::chrono::milliseconds(delay_time));
+        }
+    }
+
+    // Software Reset
+    bool DisplaySpi::sw_reset(int delay_time)
+    {
+        bool res = send_cmd(static_cast<uint8_t>(LcdCmd::SWRESET));
+        std::this_thread::sleep_for(std::chrono::milliseconds(delay_time));
+
+        return res;
     }
 
     // Send initialize sequence of commands and data to display
-    bool ILI9341::send_init_cmds()
+    bool DisplaySpi::send_init_cmds(const display_init_cmd_t* init_cmds, size_t length)
```

Comment:
There are several `res &= ...` in this function, but no checks if `res` becomes false. Does it make sense to perform additional commands once a function has failed or should the conditions have an added `res`-check, such as `for (size_t index = 0; res && index < length; index++)` ?